### PR TITLE
feat: migrate to JSpecify

### PR DESCRIPTION
**Pull request by TheMrMilchmann**
See #1008 for a detailed description.

Key changes:

- All occurrences of `javax.annotation.Nullable` have been replaced with `org.jspecify.annotations.Nullable`
  - This introduced a conceptual change as nullability effectively becomes a use-site attribute of the type. Where a method previously returned "String or null", it now returns a "nullable String". To reflect this, the location of the marker annotation was (an in a few cases had to be) changed.
  - `@Nullable public static Something.Buffer ...` becomes `public static Something.@Nullable Buffer ...`
  - The `@Nullable` marker now directly precedes the type before the `@NativeType` annotation where applicable.
- Likewise, `org.lwjgl.system.NonnullDefault` has been removed and replaced with `org.jspecify.annotations.NullMarked`.

This is a draft PR for the following reasons:

1. We need to decide whether we want to expose JSpecify as an API dependency. JSpecify recommends this, but using them as compile-only dependencies (just like we currently do with JSR305 annotations) would also work fine.
2. I'm not fully happy with the changes to the generator. If you have a better idea, please let me know.

### CODE REVIEW

File: modules/lwjgl/core/src/main/java/org/lwjgl/system/SharedLibraryLoader.java [hunk 1]
```diff
@@ -36,9 +35,7 @@ final class SharedLibraryLoader {
 
     private static final Lock EXTRACT_PATH_LOCK = new ReentrantLock();
 
-    @GuardedBy("EXTRACT_PATH_LOCK")
```
Comment (TheMrMilchmann) [inline modules/lwjgl/core/src/main/java/org/lwjgl/system/SharedLibraryLoader.java:39]: This is the only use of JSR305 annotations that can't be replaced with JSpecify. We _could_ keep JSR305 just for that but I figured it's not worth it.

**Comment by Spasi** [comment]
Great job @TheMrMilchmann, thanks!

**Comment by TheMrMilchmann** [comment]
> Is it valid syntax to have a type, the annotation, and then .Buffer?
> 
> https://github.com/LWJGL/lwjgl3/blob/2607557077ce33a35521f1fa10a79dfea050d440/modules/lwjgl/assimp/src/generated/java/org/lwjgl/assimp/AIBone.java#L235

@SWinxy This is not just valid, but the only correct syntax for the `TYPE_USE` syntax. In simple terms, these annotations must appear next to the concert type name (or `[]`) they apply to. `AIBone` is just part of the qualifier for its nested `Buffer`. This is specified in [JLS 9.7.4](https://docs.oracle.com/javase/specs/jls/se8/html/jls-9.html#jls-9.7.4).

Some examples from [JSpecify's docs](https://jspecify.dev/docs/user-guide/#type-use-annotation-syntax) (which apply to all `TYPE_USE` annotations):

> 1. For a nested static type like Map.Entry, if you want to say that the value can be `null` then the syntax is `Map.@Nullable Entry`. You can often avoid dealing with this by importing the nested type directly, but in this case `import java.util.Map.Entry` might be undesirable because `Entry` is such a common type name.
>
> 2. For an array type, if you want to say that the elements of the array can be `null` then the syntax is `@Nullable String[]`. If you want to say that the array itself can be `null` then the syntax is `String @Nullable []`. And if both the elements and the array itself can be `null`, the syntax is `@Nullable String @Nullable []`.

**Comment by SWinxy** [comment]
Wild! Thanks for the info.

**Pull request merged**